### PR TITLE
#987 Document security fix

### LINKED ISSUE #987: Security issue with Documents
**Describe the bug**
User can execute uploaded files and scripts. 
User can download malware code. Then find out a direct link to it, and then use this code in attacks on the site.

**To Reproduce**
1. Go to Documents page and upload js file with code "`alert('Hi!');`"
2. Go to activity feed and click to file. 
3. See alert "Hi!" 
4. Or click to file Documents page and see the same alert. 

**Expected behavior / Solution:**
- JS script shouldn't be executable.
- Files shouldn't be accessible for direct link. User shouldn't know or see direct link to file (see how it works in woocommece downloads).
- Files shouldn't be located in `/wp-content/uploads/` folder. It should be another folder, which can be hidden from search engines with using `robots.txt` file. For example `/wp-content/uploads/documents/` (Related #668)
- You may change default filename to random name (or name + salt). Original filename could be saved to DB. 

**Screenshots / GIF**
bp-bug

**Configuration:**
- buddyboss-platform 1.4.0.2 
- buddyboss-theme 1.4.6 and 1.5.0
- WordPress 5.4.1

### LINKED ISSUE #991: Add a limits to the number of uploaded images to the activity feed
**Is your feature request related to a problem? Please describe.**
User can upload a lot of images per one activity feed post. It can slow down the system and fill up memory quickly.
**Screenshot**
image

**Describe the solution you'd like**
Add to the platform settings an ability to set the limits of the maximum number of uploaded images to the activity feed. 

image

**Describe alternatives you've considered**
This solution is working in your [BuddyBoss Media](https://www.buddyboss.com/product/buddyboss-media/) plugin.

- buddyboss-platform 1.4.0.2

### PR DESCRIPTION
**Pull request by chetansatasiya**
### All Submissions:

* [x] Have you followed the [Pull Requests Guidelines](https://github.com/buddyboss/buddyboss-platform/wiki/Submitting-Pull-Requests#pull-request-guidelines)?
* [x] Does your code follow the [WordPress' Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Fixes #987 
Fixes #991

### How to test the changes in this Pull Request:

1.
2.
3.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

### CODE REVIEW

File: src/bp-document/bp-document-filters.php [hunk 1]
```diff
@@ -756,3 +756,489 @@ function bp_document_activity_update_document_privacy( $activity ) {
...
+add_filter( 'mod_rewrite_rules', 'bp_document_protect_download_rewite_rules' );
+
+function bp_document_check_download_folder_protection() {
+
+	$upload_dir     = wp_get_upload_dir();
+	$downloads_path = $upload_dir['basedir'] . '/bb_documents';
+	$file_path      	= $downloads_path . '/.htaccess';
+	if ( false === ( $value = get_transient( 'bp_document_transient_htaccess' ) ) || ! file_exists( $file_path ) ) {
+
+		$file_content    = 'deny from all
```
Comment (AiratHalitov) [inline src/bp-document/bp-document-filters.php:790]: This line does not allow downloading files.
Details: https://github.com/buddyboss/buddyboss-platform/issues/987#issuecomment-640288019

Comment (AiratHalitov) [inline src/bp-document/bp-document-filters.php:790]: This line does not allow downloading files.
Details: #987 (comment)

**Comment by chetansatasiya** [comment]
@AiratHalitov how you are trying to download the file? FYI: You never download the file like https://example.com/wp-content/bb_documents/2020/06/cuz.pdf

**Comment by AiratHalitov** [comment]
@chetansatasiya 
When I trying to download PDF file from direct link I see 403 error
```
Forbidden
You don't have permission to access this resource.
```

If I will delete `deny from all` line then file will be opened in my browser and then I can click CTRL+S to save. 

---
What about adding empty `index.php` file into `bb_documents` folder with code:
```
<?php
// Silence is golden.
```

**Comment by chetansatasiya** [comment]
> @chetansatasiya
> When I trying to download PDF file from direct link I see 403 error
> 
> ```
> Forbidden
> You don't have permission to access this resource.
> ```
> 
> If I will delete `deny from all` line then file will be opened in my browser and then I can click CTRL+S to save.
> 
> What about adding empty `index.php` file into `bb_documents` folder with code:
> 
> ```
> <?php
> // Silence is golden.
> ```

@AiratHalitov If I allow the direct access of file then it will be a security issues again and people will able to download the file so we are not allowing the bb_documents folder direct access and will not remove that `deny from all` from the .htaccess file otherwise it will be a hug issues and people will see the documents.

**Comment by chetansatasiya** [comment]
@AiratHalitov Now for the documents only the way is download the file OR a folder via our download link only there is no any other way you can download from the system and I think it's a strong security layer of peoples documents.

**Comment by AiratHalitov** [comment]
@chetansatasiya 
See how it works on my test server:
docs-2

I tried to use "Download" and "Copy Download Link" buttons.

**Pull request merged**